proto/v1beta1/intoto.proto

```diff
@@ -0,0 +1,145 @@
+
+import "google/protobuf/struct.proto";
+
+// This file defines the proto that matches the schema of an in-toto link
+// file. in-toto is a framework to protect a software supply chain. You can read
+// more about in-toto at https://in-toto.io/ and you can find the current
+// version of the in-toto specification at
+// https://github.com/in-toto/docs/blob/master/in-toto-spec.md
+
+
+// A signature object consists of the KeyID used and the signature itself.
+message Signature {
+  string keyid = 1 [json_name="keyid"];
+  string sig = 2 [json_name="sig"];
+}
+
+// An object defined to capture the by products of a step.
+message ByProductObject {
+  int32 return_value = 1 [json_name="return-value"];
+  string stderr = 2 [json_name="stderr"];
+  string stdout = 3 [json_name="stdout"];
+}
+
+// This corrresponds to the core in-toto link attestation.
+message Link {
+  // This field contains the full command executed for the step. This can also
+  // be empty if attestations are generated for operations that aren't directly
+  // mapped to a specific command. Each term in the command is an independent
+  // string in the list. An example of a command in the in-toto metadata field
+  // is:
+  // "command": ["git", "clone", "https://github.com/in-toto/demo-project.git"]
+  repeated string command = 1 [json_name="command"];
+
+  // Materials are the supply chain artifacts that go into the step and are used
+  // for the operation performed. The key of the map is the path of the artifact
+  // and the structure contains the recorded hash information. An example is:
+  // "materials": {
+  //   "foo/bar": {
+  //     "sha256": "ebebf..."
+  //   }
+  // }
+  map<string, google.protobuf.Struct> materials = 2 [json_name="materials"];
+
+  // Products are the supply chain artifacts generated as a result of the step.
+  // The structure is identical to that of materials.
+  map<string, google.protobuf.Struct> products = 3 [json_name="products"];
+
+  // By Products are made up of the data that is generated as part of a software
+  // supply chain step, but are not the actual result of the step.
+  ByProductObject byproducts = 4 [json_name="byproducts"];
+
+  // This is a field that can be used to capture information about the
+  // environment. It is recommended for this field to contain information that
+  // details environment variables, filesystem information, and the present
+  // working directory. The structure of this field is:
+  // "environment": {
+  //   "variables": "<ENV>",
+  //   "filesystem": "<FS>",
+  //   "workdir": "<CWD>"
+  // }
+  map<string, google.protobuf.Struct> environment = 5 [json_name="environment"];
+}
+
+// This corresponds to a full signed attestation - it is made up of one or more
+// signatures and the core link attestation itself. This is used for occurrences
+// of a Grafeas in-toto note.
+message Details {
+  repeated Signature signatures = 1 [json_name="signatures"];
+  Link link = 2 [json_name="signed"];
```

Comment:
So, a `Link` is an occurrence? and also a note?

Why are there multiple signatures and only 1 link?

Reply:
I'm going to briefly include some in-toto semantics here. in-toto has two types of metadata - [layouts](https://github.com/in-toto/docs/blob/master/in-toto-spec.md#43-file-formats-layout) and [links](https://github.com/in-toto/docs/blob/master/in-toto-spec.md#44-file-formats-namekeyid-prefixlink). A single layout is created to define the entire software supply chain, with all the keys used, the [steps](https://github.com/in-toto/docs/blob/master/in-toto-spec.md#431-steps) that make up all the different operations, and how they're 'chained' together. Links are generated when these steps are actually performed.

Now, the thinking is to have a Grafeas note that corresponds to a step in the supply chain - it identifies the step by name, the keys to be used to verify the step's link metadata, [artifact rules](https://github.com/in-toto/docs/blob/master/in-toto-spec.md#433-artifact-rules) that identify the files that were involved in the step (which are the materials and products), and an expected command which is the command that is to be used to perform this step (this can be blank). All these fields actually live in the layout of the supply chain, under the corresponding step. Every time the step is performed, we have link metadata being generated that is recorded as an occurrence.

> Why are there multiple signatures and only 1 link?

Let me verify something here.

Reply:
> Why are there multiple signatures and only 1 link?

The in-toto metaformat allows for multiple signatures because theoretically a link can have more than 1 signature. The metaformat also carries over to layouts which usually do contain 1+ signatures. While the links actually passed to the Grafeas may only contain a single signature, we need `repeated` to match the in-toto metaformat.